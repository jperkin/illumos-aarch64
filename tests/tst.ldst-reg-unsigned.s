/*
 * This file and its contents are supplied under the terms of the
 * Common Development and Distribution License ("CDDL"), version 1.0.
 * You may only use this file in accordance with the terms of version
 * 1.0 of the CDDL.
 *
 * A full copy of the text of the CDDL should have accompanied this
 * source.  A copy of the CDDL is also available via the Internet at
 * http://www.illumos.org/license/CDDL.
 */

/*
 * Copyright 2017 Joyent, Inc.
 */

/*
 * C3.3.13 Load/store register (unsigned immediate)
 *
 * C5.6.143 PRFM is part of this group but is tested in XXX.
 *
 * Test the minimum/maximum values for both registers and unsigned immediates.
 */
.macro ldst_reg_unsigned_32 op, mult=1
	\op	w0,  [x0]
	\op	wzr, [sp, #(4095*\mult)]
.endm

.macro ldst_reg_unsigned_64 op, mult=1
	\op	x0,  [x0]
	\op	xzr, [sp, #(4095*\mult)]
.endm

.macro ldst_reg_unsigned_simd op
	\op	b0,  [x0]
	\op	b31, [sp, #4095]
	\op	h0,  [x0]
	\op	h31, [sp, #8190]
	\op	s0,  [x0]
	\op	s31, [sp, #16380]
	\op	d0,  [x0]
	\op	d31, [sp, #32760]
	\op	q0,  [x0]
	\op	q31, [sp, #65520]
.endm

.text
.globl libdis_test
.type libdis_test, @function
libdis_test:
	ldst_reg_unsigned_32   strb
	ldst_reg_unsigned_32   ldrb
	ldst_reg_unsigned_64   ldrsb
	ldst_reg_unsigned_32   ldrsb
	ldst_reg_unsigned_simd str
	ldst_reg_unsigned_simd ldr
	ldst_reg_unsigned_32   strh  2
	ldst_reg_unsigned_32   ldrh  2
	ldst_reg_unsigned_64   ldrsh 2
	ldst_reg_unsigned_32   ldrsh 2
	ldst_reg_unsigned_32   str   4
	ldst_reg_unsigned_32   ldr   4
	ldst_reg_unsigned_64   ldrsw 4
	ldst_reg_unsigned_64   str   8
	ldst_reg_unsigned_64   ldr   8
.size libdis_test, [.-libdis_test]
