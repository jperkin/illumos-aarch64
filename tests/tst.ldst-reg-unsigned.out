    libdis_test:      00 00 00 39  strb w0, [x0]
    libdis_test+0x4:  ff ff 3f 39  strb wzr, [sp, #4095]
    libdis_test+0x8:  00 00 40 39  ldrb w0, [x0]
    libdis_test+0xc:  ff ff 7f 39  ldrb wzr, [sp, #4095]
    libdis_test+0x10: 00 00 80 39  ldrsb x0, [x0]
    libdis_test+0x14: ff ff bf 39  ldrsb xzr, [sp, #4095]
    libdis_test+0x18: 00 00 c0 39  ldrsb w0, [x0]
    libdis_test+0x1c: ff ff ff 39  ldrsb wzr, [sp, #4095]
    libdis_test+0x20: 00 00 00 3d  str b0, [x0]
    libdis_test+0x24: ff ff 3f 3d  str b31, [sp, #4095]
    libdis_test+0x28: 00 00 00 7d  str h0, [x0]
    libdis_test+0x2c: ff ff 3f 7d  str h31, [sp, #8190]
    libdis_test+0x30: 00 00 00 bd  str s0, [x0]
    libdis_test+0x34: ff ff 3f bd  str s31, [sp, #16380]
    libdis_test+0x38: 00 00 00 fd  str d0, [x0]
    libdis_test+0x3c: ff ff 3f fd  str d31, [sp, #32760]
    libdis_test+0x40: 00 00 80 3d  str q0, [x0]
    libdis_test+0x44: ff ff bf 3d  str q31, [sp, #65520]
    libdis_test+0x48: 00 00 40 3d  ldr b0, [x0]
    libdis_test+0x4c: ff ff 7f 3d  ldr b31, [sp, #4095]
    libdis_test+0x50: 00 00 40 7d  ldr h0, [x0]
    libdis_test+0x54: ff ff 7f 7d  ldr h31, [sp, #8190]
    libdis_test+0x58: 00 00 40 bd  ldr s0, [x0]
    libdis_test+0x5c: ff ff 7f bd  ldr s31, [sp, #16380]
    libdis_test+0x60: 00 00 40 fd  ldr d0, [x0]
    libdis_test+0x64: ff ff 7f fd  ldr d31, [sp, #32760]
    libdis_test+0x68: 00 00 c0 3d  ldr q0, [x0]
    libdis_test+0x6c: ff ff ff 3d  ldr q31, [sp, #65520]
    libdis_test+0x70: 00 00 00 79  strh w0, [x0]
    libdis_test+0x74: ff ff 3f 79  strh wzr, [sp, #8190]
    libdis_test+0x78: 00 00 40 79  ldrh w0, [x0]
    libdis_test+0x7c: ff ff 7f 79  ldrh wzr, [sp, #8190]
    libdis_test+0x80: 00 00 80 79  ldrsh x0, [x0]
    libdis_test+0x84: ff ff bf 79  ldrsh xzr, [sp, #8190]
    libdis_test+0x88: 00 00 c0 79  ldrsh w0, [x0]
    libdis_test+0x8c: ff ff ff 79  ldrsh wzr, [sp, #8190]
    libdis_test+0x90: 00 00 00 b9  str w0, [x0]
    libdis_test+0x94: ff ff 3f b9  str wzr, [sp, #16380]
    libdis_test+0x98: 00 00 40 b9  ldr w0, [x0]
    libdis_test+0x9c: ff ff 7f b9  ldr wzr, [sp, #16380]
    libdis_test+0xa0: 00 00 80 b9  ldrsw x0, [x0]
    libdis_test+0xa4: ff ff bf b9  ldrsw xzr, [sp, #16380]
    libdis_test+0xa8: 00 00 00 f9  str x0, [x0]
    libdis_test+0xac: ff ff 3f f9  str xzr, [sp, #32760]
    libdis_test+0xb0: 00 00 40 f9  ldr x0, [x0]
    libdis_test+0xb4: ff ff 7f f9  ldr xzr, [sp, #32760]
